Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h

```diff
@@ -172,6 +173,14 @@ ALWAYS_INLINE void AbstractInterpreter<AbstractStateType>::filterByType(Edge& ed
 template<typename AbstractStateType>
 void AbstractInterpreter<AbstractStateType>::verifyEdge(Node* node, Edge edge)
 {
+    if (edge.node()->isTuple()) {
+        if (edge.useKind() == UntypedUse && node->op() == ExtractFromTuple)
+            return;
+        DFG_CRASH(m_graph, node, toCString("Tuple edge verification error: ", node, "->", edge, " was expected to have Untyped use kind (had ", edge.useKind(),
+            "). Has type ", SpeculationDump(m_state.forTupleNodeWithoutFastForward(edge.node(), node->extractOffset()).m_type)).data(),
+            AbstractInterpreterInvalidType, node->op(), edge->op(), edge.useKind(), m_state.forNodeWithoutFastForward(node).m_type);
```

Comment:
In this case, the node is Extract and the edge is from Extract -> the tuple producing node, right?

```
D@1: SomethingMakesATuple(...)
D@2: Extract(D@1:UntypedUse, <<0)
D@3: GetByOffset(D@2:CellUse)
```

This case is only reached for D@2 -> D@1. D@3->D@2 would fall back to the normal validation, since D@2 has isTuple == false, right? Or am I misunderstanding how this works?

So what would it mean to speculate / prove anything about the type of the entire tuple, since our type system doesn't have tuple types?

Reply:
@kmiller68 ping